### DISCUSSION TITLE
canAccessPanel for multiple panels

### DISCUSSION THREAD

**Discussion by Ssionn** [filamentphp/filament#12838] [Help] [answered]
### Package

Panel builder

### Package Version

v.3.0.0

### How can we help you?

I'm relatively new to filament. How exactly can I give the canAccessPanel function rules for a specific panel? For example, I have two panels. One for the user and one for the admin, yet for the one below, I want to only add for the admin panel. How would I go about this?

 ```
public function canAccessPanel(Panel $panel): bool
 {
        return str_ends_with($this->email, '@example.com') && $this->hasVerifiedEmail();
 }
```

**Comment by leandrocfe** [comment, marked as answer]
use `$panel->getId()`

```php
public function canAccessPanel(Panel $panel): bool
{
    return match($panel->getId()) {
        default => false,
        'admin' => str_ends_with($this->email, '@example.com'),
    };
}
```

**Reply by bilogic** [reply]
If I have 10 panels, does `canAccessPanel` get called for every panel? Or only once during login?

**Reply by bilogic** [reply]
https://filamentphp.com/docs/3.x/panels/users#authorizing-access-to-the-panel 

1. The big security problem with this is that, it is only asserted once when the user logs in and *never again* 
2. So technically, I can login as a normal user to my `/account` panel and manually type my URL to load the `/super-admin` panel
3. Don't believe me? Try it.

**Reply by aziz-prasetyo** [reply]
> https://filamentphp.com/docs/3.x/panels/users#authorizing-access-to-the-panel
> 
> 1. The big security problem with this is that, it is only asserted once when the user logs in and _never again_
> 2. So technically, I can login as a normal user to my `/account` panel and manually type my URL to load the `/super-admin` panel
> 3. Don't believe me? Try it.

Yep, we don't trust you @bilogic. I managed to create two separate panels just by adding the following code, and access was magically restricted.

```php
public function canAccessPanel(Panel $panel): bool
{
    return $this->role->value === $panel->getId();
}

**Reply by bilogic** [reply]
Yes, best to trust your own code.

**Reply by micraux** [reply]
@bilogic, in the documentation,you can find 2 examples.

Of course, if you use the first example, all allowed users can access all panels, but the second one clearly shows how to deal with restricting access to the `admin` panel.

Just try using `if ($panel->getId() === 'super-admin')` in your example, and you should be good to go.

**Reply by bilogic** [reply]
Which examples? Show a link please. As far as I can remember, `canAccessPanel` runs only once at the point where user provides the correct password. Not whenever the user tries to access a panel.

It's ok if you can't understand this edge case. I trust my step debugging. Just remember this issue #12838 when you face a security breach.

**Reply by micraux** [reply]
Just follow the link you provided in this thread on feb 26!

And this thread is not an issue: it is just someone asking for help on how to implement.

By the way, you're wrong about `canAccessPanel` running only once, @aziz-prasetyo already told you:
- let all users access all panels, 
- log in as a non-admin user, 
- go to the admin panel,
- restrict access to admin panel for only admin users,
- reload your page, you will see the non-authorized access page.

> Don't believe me? Try it.

**Reply by bilogic** [reply]
This has devolved into a you say vs I say, as much as you think you are right, it goes both ways.

> By the way, you're wrong about `canAccessPanel` running only once

Basically you are saying `canAccessPanel` runs each time another panel access yea? Point to the code to prove it.

**Reply by danharrin** [reply]
This middleware runs on every panel request, unless you have removed it from the middleware stack in your panel provider 

https://github.com/filamentphp/filament/blob/3.x/packages/panels/src/Http/Middleware/Authenticate.php

**Reply by bilogic** [reply]
great thanks

**Reply by Asheshanghab** [reply]
But canAccesPanel only works each time when user tries to authenticate but what if the user is already logged in and want to prevent after particular time to not access panel ?
